redbot/core/modlog.py

```diff
@@ -47,10 +47,35 @@
 _CASES = "CASES"
 _SCHEMA_VERSION = 4
 
+_data_deletion_lock = asyncio.Lock()
 
 _ = Translator("ModLog", __file__)
 
 
+async def _process_data_deletion(
+    *, requester: Literal["discord_deleted_user", "owner", "user", "user_strict"], user_id: int
+):
+    if requester != "discord_deleted_user":
+        return
+
+    # Oh, how I wish it was as simple as I wanted...
+
+    key_paths = []
+
+    async with _data_deletion_lock:
+        all_cases = await _config.custom(_CASES).all()
+        async for guild_id_str, guild_cases in AsyncIter(all_cases.items(), steps=100):
+            async for case_num_str, case in AsyncIter(guild_cases.items(), steps=100):
+                if case.get("user", 0) == user_id:
+                    key_paths.append((guild_id_str, case_num_str))
+
+        async with _config.custom(_CASES).all() as all_cases:
+            for guild_id_str, case_num_str in key_paths:
+                case = all_cases[guild_id_str][case_num_str]
+                case["user"] = 0xDE1
+                case.pop("last_nown_username", None)
```

Comment:
Why can't the guild scope do something like this to reduce writes?

Reply:
"It's complicated"

I'd either need to extend config as part of this PR too or use things we can't support being done by 3rd party creators. As many creators use core cogs as a reference, I'll leave this to be improved in future work.